heron/tmaster/src/cpp/manager/tmaster.cpp

```diff
@@ -138,6 +133,68 @@ TMaster::TMaster(const std::string& _zk_hostport, const std::string& _topology_n
   }, true, PROCESS_METRICS_FREQUENCY), 0);
 }
 
+void TMaster::EstablishPackingPlan(EventLoop::Status) {
+  state_mgr_->SetPackingPlanWatch(tmaster_location_->topology_name(), [this]() {
+    this->FetchPackingPlan();
+  });
+  FetchPackingPlan();
+}
+
+void TMaster::FetchPackingPlan() {
+  auto packing_plan = new proto::system::PackingPlan();
+
+  auto cb = [packing_plan, this](proto::system::StatusCode status) {
+    this->OnPackingPlanFetch(packing_plan, status);
+  };
+
+  state_mgr_->GetPackingPlan(tmaster_location_->topology_name(), packing_plan, std::move(cb));
+}
+
+void TMaster::OnPackingPlanFetch(proto::system::PackingPlan* newPackingPlan,
+                                 proto::system::StatusCode _status) {
+  if (_status != proto::system::OK) {
+    LOG(INFO) << "PackingPlan Fetch failed with status " << _status;
+    LOG(INFO) << "Retrying after " << STATE_MANAGER_RETRY_FREQUENCY << " micro seconds ";
+    CHECK_GT(eventLoop_->registerTimer([this](EventLoop::Status) {
+      this->FetchPackingPlan();
+    }, false, STATE_MANAGER_RETRY_FREQUENCY), 0);
+  } else {
+    // We got a new PackingPlan.
+    LOG(INFO) << "Fetched PackingPlan of " << newPackingPlan->container_plans().size()
+              << " containers.";
+
+    if (!packing_plan_) {
+      LOG(INFO) << "Received initial packing plan. Initializing absent_stmgrs_";
+      packing_plan_ = newPackingPlan;
+      // We will keep the list of stmgrs with us.
+      // In case a assignment already exists, we will throw this
+      // list out and re-init with whats in assignment.
+      absent_stmgrs_.clear();
+      for (sp_uint32 i = 0; i < packing_plan_->container_plans_size(); ++i) {
+        LOG(INFO) << "Adding container id " << packing_plan_->container_plans(i).id()
+                  << " to absent_stmgrs_";
+        // the packing plan represents container ids by the numerical id of the container. The
+        // physical plan prepends "stmgr-" to the integer and represents it as a string.
+        absent_stmgrs_.insert("stmgr-" + std::to_string(packing_plan_->container_plans(i).id()));
+      }
+
+      // chain TMaster location setting to the completion of packing plan info above
+      EstablishTMaster(EventLoop::TIMEOUT_EVENT);
+    } else {
+      if (packing_plan_ != newPackingPlan) {
+        LOG(INFO) << "Packing plan changed. Deleting physical plan and restarting TMaster to "
+                  << "reset internal state. Exiting";
+        state_mgr_->DeletePhysicalPlan(tmaster_location_->topology_name(),
```

Comment:
No, we're not actually. What is the deterministic way to know that? We could set a flag during a successful `SetTMasterLocationDone` call to indicate that we're TMaster since we successfully set our location.